services/provider/server/cephblockpool.go

```diff
@@ -0,0 +1,126 @@
+package server
+
+import (
+	"context"
+	"crypto/md5"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"slices"
+
+	rookCephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+)
+
+type cephBlockPoolManager struct {
+	client    client.Client
+	namespace string
+}
+
+func newCephBlockPoolManager(cl client.Client, namespace string) (*cephBlockPoolManager, error) {
+	return &cephBlockPoolManager{
+		client:    cl,
+		namespace: namespace,
+	}, nil
+}
+
+func (c *cephBlockPoolManager) EnableBlockPoolMirroring(ctx context.Context, blockPoolName string) error {
+
+	cephBlockPool, err := c.get(ctx, blockPoolName)
+	if err != nil {
+		return err
+	}
+
+	_, err = ctrl.CreateOrUpdate(ctx, c.client, cephBlockPool, func() error {
+		cephBlockPool.Spec.Mirroring.Enabled = true
+		cephBlockPool.Spec.Mirroring.Mode = "image"
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("failed to enable mirroring on CephBlockPool resource with name %q. %v", blockPoolName, err)
+	}
+
+	return nil
+
+}
+
+func (c *cephBlockPoolManager) SetBootstrapSecretRef(ctx context.Context, blockPoolName string, secretData map[string][]byte) error {
+
+	cephBlockPool, err := c.get(ctx, blockPoolName)
+	if err != nil {
+		return err
+	}
+
+	// create the secret
+	secretJSON, err := json.Marshal(secretData)
+	if err != nil {
+		return err
+	}
+	secretNameMd5Sum := md5.Sum(secretJSON)
+	secretName := hex.EncodeToString(secretNameMd5Sum[:16])
+
+	gvk, err := apiutil.GVKForObject(cephBlockPool, c.client.Scheme())
+	if err != nil {
+		return fmt.Errorf("failed to get gvk for cephBlockPool %q. %w", cephBlockPool.Name, err)
+	}
```

Comment:
I don't think you need to extract this info, doesn't controller-runtime has utils for this? there should be some `SetControllerReference` 🤔